tests/Shaolinq.Tests/DataAccessModelHookTests.cs

```diff
@@ -0,0 +1,406 @@
+				if (complete) scope.Complete();
+			}
+
+			Assert.AreEqual(complete ? 1 : 0, this.testDataModelHook.CommitCount);
+			Assert.AreEqual(complete || !flush ? 0 : 1, this.testDataModelHook.RollbackCount);
+			Assert.AreEqual(complete ? 1 : 0, hook2.CommitCount);
+			Assert.AreEqual(complete || !flush ? 0 : 1, hook2.RollbackCount);
+		}
+
+		[TestCase(false, false)]
+		[TestCase(false, true)]
+		[TestCase(true, false)]
+		[TestCase(true, true)]
+		public void Test_Nested_DataAccessScope_Inner_Complete_Calls_DataModelHook(bool flush, bool complete)
+		{
+			using (var outerScope = new DataAccessScope())
+			{
+				var cat1 = this.model.Cats.Create();
+
+				using (var innerScope = new DataAccessScope())
+				{
+					var cat2 = this.model.Cats.Create();
+
+					innerScope.Complete();
+				}
+
+				var cat3 = this.model.Cats.Create();
+
+				if (flush) outerScope.Flush();
+				if (complete) outerScope.Complete();
+			}
+
+			Assert.AreEqual(complete ? 1 : 0, this.testDataModelHook.CommitCount);
+			Assert.AreEqual(complete || !flush ? 0 : 1, this.testDataModelHook.RollbackCount);
+		}
+
+		[Test]
+		public void Test_Nested_DataAccessScope_Inner_Not_Complete_Should_Throw_TransactionAbortedException()
+		{
+			using (var outerScope = new DataAccessScope())
+			{
+				Console.WriteLine(DataAccessTransaction.Current.SystemTransaction?.TransactionInformation.LocalIdentifier);
+
+				var cat1 = this.model.Cats.Create();
+
+				using (var innerScope = new DataAccessScope())
+				{
+					var cat2 = this.model.Cats.Create();
+				}
+
+				Assert.Throws<TransactionAbortedException>(() =>
+				{
+					var cat3 = this.model.Cats.Create();
+				});
+			}
+		}
+
+		[TestCase(false, false)]
+		[TestCase(false, true)]
+		[TestCase(true, false)]
+		[TestCase(true, true)]
+		public void Test_Nested_TransactionScope_Inner_Complete_Calls_DataModelHook(bool flush, bool complete)
+		{
+			using (var outerScope = new TransactionScope())
+			{
+				var cat1 = this.model.Cats.Create();
+
+				using (var innerScope = new TransactionScope())
+				{
+					var cat2 = this.model.Cats.Create();
+
+					innerScope.Complete();
+				}
+
+				var cat3 = this.model.Cats.Create();
+
+				if (flush) outerScope.Flush();
+				if (complete) outerScope.Complete();
+			}
+
+			Assert.AreEqual(complete ? 1 : 0, this.testDataModelHook.CommitCount);
+			Assert.AreEqual(complete || !flush ? 0 : 1, this.testDataModelHook.RollbackCount);
+		}
+
+		[Test]
+		public void Test_Nested_TransactionScope_Inner_Not_Complete_Should_Throw_TransactionAbortedException()
+		{
+			using (var outerScope = new TransactionScope())
+			{
+				var cat1 = this.model.Cats.Create();
+
+				using (var innerScope = new TransactionScope())
+				{
+					var cat2 = this.model.Cats.Create();
+				}
+
+				Assert.Throws<TransactionAbortedException>(() =>
+				{
+					var cat3 = this.model.Cats.Create();
+				});
+			}
+		}
+
+
+		[TestCase(false, false)]
+		[TestCase(false, true)]
+		[TestCase(true, false, Ignore = "Hangs")]
```

Comment:
This test is a bit of an edge case, but this combination causes the test to hang completely.